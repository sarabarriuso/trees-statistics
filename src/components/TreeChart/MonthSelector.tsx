import Slider from '@mui/material/Slider';

interface IMonthSelectorProps {
  description: string;
  defaultMonth: number;
  monthSelectedCallback: (value: number) => void;
}
const MonthSelector: React.FC<IMonthSelectorProps> = ({
  description,
  defaultMonth,
  monthSelectedCallback,
}) => {
  const DEFAULT_CLASSNAME = 'month-selector';

  const marks = [
    {
      value: 1,
      label: 'Jan',
    },
    {
      value: 2,
      label: 'Feb',
    },
    {
      value: 3,
      label: 'Mar',
    },
    {
      value: 4,
      label: 'Apr',
    },
    {
      value: 5,
      label: 'May',
    },
    {
      value: 6,
      label: 'Jun',
    },
    {
      value: 7,
      label: 'Jul',
    },
    {
      value: 8,
      label: 'Aug',
    },
    {
      value: 9,
      label: 'Sep',
    },
    {
      value: 10,
      label: 'Oct',
    },
    {
      value: 11,
      label: 'Nov',
    },
    {
      value: 12,
      label: 'Dec',
    },
  ];

  const handleChange = (
    _event: Event | React.SyntheticEvent<Element, Event>,
    value: number | number[],
  ) => {
    monthSelectedCallback(value as number);
  };

  return (
    <div className={DEFAULT_CLASSNAME}>
      <p>{description}</p>
      <Slider
        aria-label='Day'
        defaultValue={defaultMonth}
        track={false}
        valueLabelDisplay='auto'
        step={1}
        marks={marks}
        min={1}
        max={12}
        onChangeCommitted={handleChange}
      />
    </div>
  );
};

export default MonthSelector;
