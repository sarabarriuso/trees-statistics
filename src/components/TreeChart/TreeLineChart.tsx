import { Line } from 'react-chartjs-2';
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from 'chart.js';

import { IPlantedTreeData } from '../../interfaces/IPlantedTreeData';

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
);

interface ITreeBarChartProps {
  chartData: Array<IPlantedTreeData>;
}

const TreeLineChart: React.FC<ITreeBarChartProps> = ({ chartData }) => {
  const DEFAULT_CLASSNAME = 'tree-bar-chart';
  const graphData = {
    labels: chartData.map(
      (dataPoint) => dataPoint.date.toISOString().split('T')[0],
    ),
    datasets: [
      {
        label: 'Trees planted',
        data: chartData.map((dataPoint) => dataPoint.count),
        backgroundColor: '#1976d2',
      },
    ],
  };

  return (
    <div className={DEFAULT_CLASSNAME}>
      <Line
        options={{
          responsive: true,
          plugins: {
            legend: {
              position: 'bottom',
            },
            title: {
              display: true,
              text: 'Trees planted per day',
            },
          },
        }}
        data={graphData}
      />
    </div>
  );
};

export default TreeLineChart;
