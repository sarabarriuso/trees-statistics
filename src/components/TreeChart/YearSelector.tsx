import Slider from '@mui/material/Slider';
interface IMonthSelectorProps {
  description: string;
  defaultYear: number;
  minYear?: number;
  yearSelectedCallback: (value: number) => void;
}
const MonthSelector: React.FC<IMonthSelectorProps> = ({
  description,
  defaultYear,
  minYear,
  yearSelectedCallback,
}) => {
  const DEFAULT_CLASSNAME = 'year-selector';

  const handleChange = (
    _event: Event | React.SyntheticEvent<Element, Event>,
    value: number | number[],
  ) => {
    yearSelectedCallback(value as number);
  };

  const marks = [];
  for (let i = 2019; i <= new Date().getFullYear(); i++) {
    marks.push({ value: i, label: i.toString() });
  }

  return (
    <div className={DEFAULT_CLASSNAME}>
      <p>{description}</p>
      <Slider
        aria-label='Day'
        defaultValue={defaultYear}
        track={false}
        valueLabelDisplay='auto'
        step={1}
        marks={marks}
        min={minYear || 2019}
        max={new Date().getFullYear()}
        onChangeCommitted={handleChange}
      />
    </div>
  );
};

export default MonthSelector;
