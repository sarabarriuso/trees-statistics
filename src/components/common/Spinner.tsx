import React from 'react';
import { ReactComponent as SpinnerSvg } from '../../assets/images/spinner.svg';

export interface ISpinnerProps {
  caption: string;
}

const Spinner: React.FC<ISpinnerProps> = ({ caption }) => {
  const DEFAULT_CLASSNAME = 'spinner';

  return (
    <div className={DEFAULT_CLASSNAME}>
      <div className={`${DEFAULT_CLASSNAME}__wrap`}>
        <SpinnerSvg className={`${DEFAULT_CLASSNAME}__svg`} />
        <p>{caption}</p>
      </div>
    </div>
  );
};

export default Spinner;
