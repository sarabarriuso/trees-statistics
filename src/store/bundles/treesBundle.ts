import { IPlantedTreeData } from '../../interfaces/IPlantedTreeData';

export enum ActionTypes {
  SET_PLANTED_TREES_DATA = 'TREES_BUNDLE_SET_PLANTED_TREES_DATA',
  SET_LOADING_PLANTED_TREES_DATA = 'TREES_BUNDLE_SET_LOADING_PLANTED_TREES_DATA',
}

interface ISetPlantedTreesInfo {
  type: ActionTypes.SET_PLANTED_TREES_DATA;
  plantedTreesInfo: Array<IPlantedTreeData> | undefined;
}
interface ISetLoadingPlantedTreesInfo {
  type: ActionTypes.SET_LOADING_PLANTED_TREES_DATA;
  loadingPlantedTreesInfo: boolean;
}

// Action Combinator
type Action = ISetPlantedTreesInfo | ISetLoadingPlantedTreesInfo;

// State Slice Definition
export interface ITreesState {
  loadingPlantedTreesData: boolean;
  plantedTreesData: Array<IPlantedTreeData> | undefined;
}

// Action Creators
export const actionCreators = {
  setLoadingPlantedTreesData(
    loadingPlantedTreesInfo: boolean,
  ): ISetLoadingPlantedTreesInfo {
    return {
      type: ActionTypes.SET_LOADING_PLANTED_TREES_DATA,
      loadingPlantedTreesInfo,
    };
  },
  setPlantedTreesData(
    plantedTreesInfo: Array<IPlantedTreeData>,
  ): ISetPlantedTreesInfo {
    return {
      type: ActionTypes.SET_PLANTED_TREES_DATA,
      plantedTreesInfo,
    };
  },
};

// Sub-Reducers
function setPlantedTreesDataAction(
  state: ITreesState = getDefault(),
  action: ISetPlantedTreesInfo,
): ITreesState {
  return {
    ...state,
    plantedTreesData: action.plantedTreesInfo,
  };
}

function setLoadingPlantedTreesDataAction(
  state: ITreesState = getDefault(),
  action: ISetLoadingPlantedTreesInfo,
): ITreesState {
  return {
    ...state,
    loadingPlantedTreesData: action.loadingPlantedTreesInfo,
  };
}

// Get Default
export function getDefault(): ITreesState {
  return {
    loadingPlantedTreesData: false,
    plantedTreesData: undefined,
  };
}

export default function treesReducer(
  state: ITreesState = getDefault(),
  action: Action,
): ITreesState {
  switch (action.type) {
    case ActionTypes.SET_LOADING_PLANTED_TREES_DATA:
      return setLoadingPlantedTreesDataAction(state, action);
    case ActionTypes.SET_PLANTED_TREES_DATA:
      return setPlantedTreesDataAction(state, action);
    default:
      return state;
  }
}
