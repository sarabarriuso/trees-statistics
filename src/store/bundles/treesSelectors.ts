import { IPlantedTreeData } from '../../interfaces/IPlantedTreeData';
import IState from '../state';
import { useStateSelector } from '../useSelector';

export const GetPlantedTreesData = (): Array<IPlantedTreeData> | undefined =>
  useStateSelector((state: IState) => state.treesState.plantedTreesData);
export const GetLoadingPlantedTreesData = (): boolean =>
  useStateSelector((state: IState) => state.treesState.loadingPlantedTreesData);
