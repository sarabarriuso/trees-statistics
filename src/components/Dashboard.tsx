import { GetLoadingPlantedTreesData } from '../store/bundles/treesSelectors';
import Spinner from './common/Spinner';
import Header from './Header';
import TreeChart from './TreeChart';
import t from '../language';

const Dashboard: React.FC = () => {
  const DEFAULT_CLASSNAME = 'dashboard';

  const loadingData = GetLoadingPlantedTreesData();

  return (
    <div className={DEFAULT_CLASSNAME}>
      <Header />
      {loadingData ? (
        <Spinner caption={t.text('app.dashboard.loadingData')} />
      ) : (
        <TreeChart />
      )}
    </div>
  );
};

export default Dashboard;
