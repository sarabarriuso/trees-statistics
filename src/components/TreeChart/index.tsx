import React, { useState } from 'react';
import { GetPlantedTreesData } from '../../store/bundles/treesSelectors';
import { filterChartData } from '../../utilities/treeChartHelper';
import MonthSelector from './MonthSelector';
import TreeLineChart from './TreeLineChart';
import YearSelector from './YearSelector';

const TreeChart: React.FC = () => {
  const DEFAULT_CLASSNAME = 'tree-chart';

  const [fromMonth, setFromMonth] = useState<number>(1);
  const [toMonth, setToMonth] = useState<number>(1);
  const [fromYear, setFromYear] = useState<number>(2019);
  const [toYear, setToYear] = useState<number>(1);

  const treesData = GetPlantedTreesData();

  //const treesData = useMemo(() => GetPlantedTreesData(), []);
  const treeChartData = filterChartData(
    new Date(fromYear, fromMonth - 1),
    new Date(toYear, toMonth - 1),
    treesData || [],
  );

  return (
    <div className={DEFAULT_CLASSNAME}>
      <div className={`${DEFAULT_CLASSNAME}__filters`}>
        <YearSelector
          description={'From year:'}
          defaultYear={fromYear}
          yearSelectedCallback={(value: number) => setFromYear(value)}
        />
        <MonthSelector
          description={'From month:'}
          defaultMonth={fromMonth}
          monthSelectedCallback={(value: number) => setFromMonth(value)}
        />
        <YearSelector
          description={'To year:'}
          defaultYear={toYear}
          minYear={fromYear}
          yearSelectedCallback={(value: number) => setToYear(value)}
        />
        <MonthSelector
          description={'To month:'}
          defaultMonth={toMonth}
          monthSelectedCallback={(value: number) => setToMonth(value)}
        />
      </div>

      {treesData && <TreeLineChart chartData={treeChartData} />}
    </div>
  );
};

export default TreeChart;
