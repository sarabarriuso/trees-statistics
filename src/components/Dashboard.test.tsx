import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dashboard from './Dashboard';

import * as TreesSelectors from '../store/bundles/treesSelectors';
import TreeChart from './TreeChart';
import Spinner from './common/Spinner';

export default describe('Dashboard Tests', () => {
  const getLoadingPlantedTreesDataMock = jest.spyOn(
    TreesSelectors,
    'GetLoadingPlantedTreesData',
  );

  afterEach(() => {
    getLoadingPlantedTreesDataMock.mockClear();
  });

  it('Renders successfully when is not loading tree data', () => {
    // Arrange
    getLoadingPlantedTreesDataMock.mockReturnValue(false);
    // Act
    const tree = shallow(<Dashboard />);

    //Assert
    expect(toJson(tree)).toMatchSnapshot();
    const treeChart = tree.find(TreeChart);
    const spinner = tree.find(Spinner);
    expect(treeChart.length).toBe(1);
    expect(spinner.length).toBe(0);
  });

  it('Renders successfully when is loading tree data', () => {
    // Arrange
    getLoadingPlantedTreesDataMock.mockReturnValue(true);
    // Act
    const tree = shallow(<Dashboard />);

    //Assert
    expect(toJson(tree)).toMatchSnapshot();
    const treeChart = tree.find(TreeChart);
    const spinner = tree.find(Spinner);
    expect(treeChart.length).toBe(0);
    expect(spinner.length).toBe(1);
  });
});
