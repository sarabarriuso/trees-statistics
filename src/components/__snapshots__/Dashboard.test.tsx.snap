// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Tests Renders successfully when is loading tree data 1`] = `
<div
  className="dashboard"
>
  <Header />
  <Spinner
    caption="Counting trees..."
  />
</div>
`;

exports[`Dashboard Tests Renders successfully when is not loading tree data 1`] = `
<div
  className="dashboard"
>
  <Header />
  <TreeChart />
</div>
`;
